src/bp/ui-studio/src/web/views/FlowBuilder/sidePanel/Toolbar.tsx

```diff
@@ -0,0 +1,86 @@
+import { AnchorButton, Icon, Intent, Popover, Position, Tag, Tooltip } from '@blueprintjs/core'
+import _ from 'lodash'
+import React from 'react'
+import { connect } from 'react-redux'
+import { flowEditorRedo, flowEditorUndo, saveAllFlows } from '~/actions'
+import { LeftToolbarButtons, RightToolbarButtons, Toolbar } from '~/components/Shared/Interface'
+import { canFlowRedo, canFlowUndo } from '~/reducers'
+import { getCurrentFlow } from '~/reducers'
+
+import style from './style.scss'
+
+const FlowProblems = props => {
+  const highlightNode = node => {
+    // @ts-ignore
+    window.highlightNode(props.currentFlow && props.currentFlow.name, node)
+  }
+  const hasProblems = !!props.flowProblems.length
+
+  return (
+    <Popover>
+      <Tooltip
+        content={
+          hasProblems ? <span>There are some problems with your flow.<br />Click for more details</span> : <span>No problem with your flow.</span>
+        }
+        position={Position.BOTTOM}
+      >
+        {hasProblems ? (
+          <Tag icon="error" className={style.flowProblems} intent={Intent.DANGER}>
+            {props.flowProblems.length}
+          </Tag>
+        ) : (
+          <Icon icon="info-sign" color="gray" style={{ padding: '5px 5px 0 0' }} />
+        )}
+      </Tooltip>
+      <div style={{ padding: 10 }}>
+        {props.flowProblems.map(node => (
+          <div key={node.nodeName}>
+            <a onClick={() => highlightNode(node.nodeName)}>
+              <strong>{node.nodeName}</strong>
+            </a>
+            : Missing <strong>{node.missingPorts}</strong> links
+          </div>
+        ))}
+      </div>
+    </Popover>
+  )
+}
+
+const MiniToolbar = props => {
+  return (
+    <Toolbar>
+      <LeftToolbarButtons>
```

Comment:
missing permission checkers ?

Reply:
Technically no, if you don'T have permission -> you are read only -> can't make change which needs undo